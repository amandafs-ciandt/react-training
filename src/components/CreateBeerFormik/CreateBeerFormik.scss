.create-beer-formik {
  padding: 2rem;

  &__title {
    text-align: center;
    color: #595959;
    margin-top: 0;
  }

  &__form {
    display: flex;
    flex-direction: column;
  }

  &__input-container {
    margin-bottom: 1rem;
  }

  .MuiFormControl-root,
  Button {
    width: 100%;
  }
}
