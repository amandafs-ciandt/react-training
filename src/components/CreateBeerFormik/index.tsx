import React from 'react';
import { useFormik } from 'formik';
import * as yup from 'yup';

import {
  TextField,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  FormControlLabel,
  Checkbox,
  Button,
} from '@material-ui/core';

import './CreateBeerFormik.scss';

interface IProps {
  notify: Function;
}

const beerSchema = yup.object().shape({
  beerName: yup.string().required(),
  beerType: yup.string().required(),
  hasCorn: yup.boolean(),
  ingredients: yup.string().required(),
});

const defaultValues = {
  beerName: '',
  beerType: '',
  hasCorn: false,
  ingredients: '',
};

const CreateBeerFormik = ({ notify }: IProps) => {
  const beerForm = useFormik({
    initialValues: defaultValues,
    validationSchema: beerSchema,
    onSubmit: () => {
      notify();
    },
  });

  return (
    <div className='create-beer-formik'>
      <h1 className='create-beer-formik__title'>Beer Formik</h1>
      <form
        onSubmit={beerForm.handleSubmit}
        className='create-beer-formik__form'>
        <div>
          <div className='create-beer-formik__input-container'>
            <TextField
              label='Beer name'
              id='beerName'
              name='beerName'
              variant='outlined'
              value={beerForm.values.beerName}
              onChange={beerForm.handleChange}
            />
          </div>
          <div className='create-beer-formik__input-container'>
            <FormControl variant='outlined'>
              <InputLabel id='beerTypeLabel'>Beer type</InputLabel>
              <Select
                labelId='beerTypeLabel'
                id='beerType'
                label='Beer type'
                value={beerForm.values.beerType}
                onChange={beerForm.handleChange}
                inputProps={{
                  name: 'beerType',
                  id: 'beerType',
                }}>
                <MenuItem data-testid='beer-type-option' value='ale'>
                  Ale
                </MenuItem>
                <MenuItem data-testid='beer-type-option' value='lager'>
                  Lager
                </MenuItem>
                <MenuItem data-testid='beer-type-option' value='malt'>
                  Malt
                </MenuItem>
                <MenuItem data-testid='beer-type-option' value='stout'>
                  Stout
                </MenuItem>
              </Select>
            </FormControl>
          </div>
          <div className='create-beer-formik__input-container'>
            <FormControlLabel
              control={<Checkbox color='primary' />}
              label='Has corn'
              labelPlacement='end'
              name='hasCorn'
              id='hasCorn'
              checked={beerForm.values.hasCorn}
              onChange={() =>
                beerForm.setFieldValue('hasCorn', !beerForm.values.hasCorn)
              }
            />
          </div>
          <div className='create-beer-formik__input-container'>
            <TextField
              id='ingredients'
              label='Ingredients'
              multiline
              rows={3}
              variant='outlined'
              name='ingredients'
              value={beerForm.values.ingredients}
              onChange={beerForm.handleChange}
            />
          </div>
          <Button
            disabled={!(beerForm.isValid && beerForm.dirty)}
            data-testid='form-submit-button'
            type='submit'
            variant='contained'
            color='secondary'>
            Submit
          </Button>
        </div>
      </form>
    </div>
  );
};

export default CreateBeerFormik;
