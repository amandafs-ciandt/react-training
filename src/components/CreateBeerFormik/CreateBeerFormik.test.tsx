import React from 'react';
import { shallow, ShallowWrapper, ReactWrapper } from 'enzyme';

import CreateBeerFormik from '.';

const mockedNotify = jest.fn();

const setUpShallowRendering = (): ShallowWrapper => {
  return shallow(<CreateBeerFormik notify={mockedNotify} />);
};

describe('CreateBeerFormik', () => {
  let createBeerFormikComponent: ShallowWrapper | ReactWrapper;

  beforeEach(() => {
    createBeerFormikComponent = setUpShallowRendering();
  });

  it('should render correctly', () => {
    expect(createBeerFormikComponent).toMatchSnapshot();
  });

  it('should render title', () => {
    const expectedTitle = 'Beer Formik';
    const title = createBeerFormikComponent
      .find('.create-beer-formik__title')
      .text();

    expect(title).toBe(expectedTitle);
  });

  it('should render the form', () => {
    const formExists = createBeerFormikComponent.exists(
      '.create-beer-formik__form'
    );

    expect(formExists).toBeTruthy();
  });

  it('should render the inputs', () => {
    const numberOfInputs = 4;
    const formInputs = createBeerFormikComponent.find(
      '.create-beer-formik__input-container'
    );

    expect(formInputs.length).toEqual(numberOfInputs);
  });

  it('should have submit button disabled after form is rendered', () => {
    const submitButton = createBeerFormikComponent.find(
      '[data-testid="form-submit-button"]'
    );

    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should update beerName when it is changed', () => {
    createBeerFormikComponent.find('#beerName').simulate('change', {
      persist: () => {},
      target: {
        name: 'beerName',
        value: 'Budweiser',
      },
    });

    const newValue = createBeerFormikComponent.find('#beerName').props().value;

    expect(newValue).toEqual('Budweiser');
  });

  it('should update beerType when it is changed', () => {
    createBeerFormikComponent.find('#beerType').simulate('change', {
      persist: () => {},
      target: {
        name: 'beerType',
        value: 'lager',
      },
    });

    const newValue = createBeerFormikComponent.find('#beerType').props().value;

    expect(newValue).toEqual('lager');
  });

  it('should update hasCorn when it is changed', () => {
    createBeerFormikComponent.find('#hasCorn').simulate('change', {
      persist: () => {},
      target: {
        name: 'hasCorn',
        checked: true,
      },
    });

    const newValue = createBeerFormikComponent.find('#hasCorn').props().checked;

    expect(newValue).toEqual(true);
  });

  it('should update ingredients when it is changed', () => {
    const ingredientsValue = 'Water, barley malt, rice, yeast and hops';

    createBeerFormikComponent.find('#ingredients').simulate('change', {
      persist: () => {},
      target: {
        name: 'ingredients',
        value: ingredientsValue,
      },
    });

    const newValue = createBeerFormikComponent.find('#ingredients').props()
      .value;

    expect(newValue).toEqual(ingredientsValue);
  });

  it('should submit form correctly', () => {
    //TODO
  });
});
