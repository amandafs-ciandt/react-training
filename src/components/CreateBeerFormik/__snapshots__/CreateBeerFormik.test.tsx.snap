// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBeerFormik should render correctly 1`] = `ShallowWrapper {}`;
